bsmetadata/paragraph_by_metadata_html.py

```diff
@@ -0,0 +1,87 @@
+import re
+from itertools import zip_longest
+from operator import itemgetter
+from typing import Any, Dict, List
+
+
+TEXT_CONTENT_ELEMENTS = {
+    "p",
+    "div",
+    "li",
+    "ul",
+    "ol",
+    "hr",
+    "pre",
+    "blockquote",
+    "dt",
+    "dd",
+    "dl",
+    "menu",
+    "figcaption",
+    "figure",
+}
+CONTENT_SECTIONING_ELEMENTS = {  # except "header" and "footer"
+    "h3",
+    "h2",
+    "h1",
+    "h4",
+    "h5",
+    "h6",
+    "nav",
+    "article",
+    "main",
+    "section",
+    "address",
+    "aside",
+}
+MARKERS = TEXT_CONTENT_ELEMENTS | CONTENT_SECTIONING_ELEMENTS
+
+
+def _pairify(iterable):
+    return (pair for pair in (zip_longest(*[iter(iterable)] * 2, fillvalue="")))
+
+
+def _split_by_double_lf(text: str):
+    return list("".join(pair) for pair in _pairify(re.split("(\n\n)", text)))
+
+
+def _extract_paragraphs(text_chunk: str, marker: str, char_start_idx: int):
+    paragraphs = []
+    txt_segs = list(filter(None, _split_by_double_lf(text_chunk)))
+    final_marker = f"{marker}+lf" if len(txt_segs) > 1 else marker
+    offset = 0
+    for txt_seg in txt_segs:
+        txt_seg_len = len(txt_seg)
+        paragraph = {
+            "char_end_idx": char_start_idx + offset + txt_seg_len,
+            "char_start_idx": char_start_idx + offset,
+            "key": "paragraph",
+            "type": "local",
+            "value": txt_seg,
+            "marker": final_marker,
+        }
+        paragraphs.append(paragraph)
+        offset += txt_seg_len
+    return paragraphs
+
+
+def get_paragraphs(html_metadata: List[Dict[str, Any]], text: str) -> List[Dict[str, Any]]:
+    paragraphs = []
+
+    tags = sorted(
+        [tag for tag in html_metadata if tag["value"] in MARKERS],
+        key=itemgetter("char_end_idx", "char_start_idx"),
+    )
+
+    char_start_idx = 0
+    for tag in tags:
+        char_end_idx = tag["char_end_idx"] + 1 if "p" == tag["value"] else tag["char_end_idx"]
+        txt_blk = text[char_start_idx:char_end_idx]
+        if txt_blk.strip():
```

Comment:
This is an edge case but sometimes a block contains only "`\n`" that may not be in the original text.